runtime/gc_base/ObjectAccessBarrier.cpp

```diff
@@ -1393,52 +1393,72 @@ BOOLEAN
 MM_ObjectAccessBarrier::structuralCompareFlattenedObjects(J9VMThread *vmThread, J9Class *valueClass, j9object_t lhsObject, j9object_t rhsObject, UDATA startOffset)
 {
 	bool result = true;
-	const UDATA *descriptionPtr = (UDATA *) valueClass->instanceDescription;
-	UDATA descriptionBits = 0;
 	bool const compressed = J9VMTHREAD_COMPRESS_OBJECT_REFERENCES(vmThread);
 	UDATA const referenceSize = J9VMTHREAD_REFERENCE_SIZE(vmThread);
+	bool hasReferences = J9CLASS_HAS_REFERENCES(valueClass);
+	UDATA limit = valueClass->totalInstanceSize;
+	U_32 firstFieldOffset = (U_32) valueClass->backfillOffset;
+	UDATA offset = 0;
 
 	Assert_MM_true(J9_IS_J9CLASS_VALUETYPE(valueClass));
 
-	if(((UDATA)descriptionPtr) & 1) {
-		descriptionBits = ((UDATA)descriptionPtr) >> 1;
-	} else {
-		descriptionBits = *descriptionPtr++;
-	}
+	if (hasReferences) {
+		UDATA descriptionIndex = J9_OBJECT_DESCRIPTION_SIZE - 1;
+		const UDATA *descriptionPtr = (UDATA *) valueClass->instanceDescription;
+		UDATA descriptionBits = 0;
 
-	UDATA descriptionIndex = J9_OBJECT_DESCRIPTION_SIZE - 1;
-	UDATA offset = 0;
-	UDATA limit = valueClass->totalInstanceSize;
+		if (0 != firstFieldOffset) {
+			/* subtract padding */
+			offset += firstFieldOffset;
+			descriptionBits >>= 1;
+			descriptionIndex -= 1;
+		}
 
-	U_32 firstFieldOffset = (U_32) valueClass->backfillOffset;
-	if (0 != firstFieldOffset) {
-		/* subtract padding */
-		offset += firstFieldOffset;
-		descriptionBits >>= 1;
-		descriptionIndex -= 1;
-	}
+		if (((UDATA)descriptionPtr) & 1) {
+			descriptionBits = ((UDATA)descriptionPtr) >> 1;
+		} else {
+			descriptionBits = *descriptionPtr++;
+		}
 
-	while (offset < limit) {
-		/* Determine if the slot contains an object pointer or not */
-		if(descriptionBits & 1) {
-			if (mixedObjectReadObject(vmThread, lhsObject, startOffset + offset, false) != mixedObjectReadObject(vmThread, rhsObject, startOffset + offset, false)) {
-				result = false;
-				break;
+
+		while (offset < limit) {
+			/* Determine if the slot contains an object pointer or not */
+			if (descriptionBits & 1) {
+				if (mixedObjectReadObject(vmThread, lhsObject, startOffset + offset, false) != mixedObjectReadObject(vmThread, rhsObject, startOffset + offset, false)) {
+					result = false;
+					break;
+				}
+			} else {
+				fomrobject_t lhsValue = GC_SlotObject::readSlot((fomrobject_t*)((UDATA)lhsObject + startOffset + offset), compressed);
+				fomrobject_t rhsValue = GC_SlotObject::readSlot((fomrobject_t*)((UDATA)rhsObject + startOffset + offset), compressed);
```

Comment:
this is more a question for my benefit: if these are not object pointers, what are they (and still care if it's compressed or not)?

Reply:
These are primitive fields. We still need slot sized loads because each iteration needs to match with the instance description bits. Whether we are compressed or not will determine how bytes each description bit represents.